src/helm/proxy/tokenizers/cachable_tokenizer.py

```diff
@@ -0,0 +1,183 @@
+from abc import abstractmethod
+from dataclasses import asdict
+from typing import Any, Dict, List, Optional
+
+from helm.common.cache import Cache, CacheConfig
+from helm.common.request import wrap_request_time
+from helm.common.tokenization_request import (
+    TokenizationRequest,
+    TokenizationRequestResult,
+    DecodeRequest,
+    DecodeRequestResult,
+    TokenizationToken,
+)
+from .tokenizer import Tokenizer
+
+
+class CachableTokenizer(Tokenizer):
+    def __init__(self, cache_config: CacheConfig) -> None:
+        self.cache = Cache(cache_config)
+
+    def _get_tokenizer_name(self, tokenizer: str) -> str:
+        return tokenizer.split("/")[-1]
+
+    @property
+    def use_encode_in_cache_key(self) -> bool:
+        """Whether to use the `encode` parameter in the cache key.
+
+        This is a small optimization as some tokenizers directly compute both
+        the token strings and the token ids, so the `encode` parameter is not
+        used in the tokenization logic.
+        """
+        return True
+
+    def _post_process_tokenization(
+        self, tokens: List[TokenizationToken], request: TokenizationRequest
+    ) -> List[TokenizationToken]:
+        """Post-processes the tokens before returning them.
+
+        This method is called after tokenization and after caching.
+        It is useful to make this class more modular as some tokenizers need
+        special post-processing.
+        """
+        return tokens
+
+    def _post_process_decoding(self, text: str, request: DecodeRequest) -> str:
+        """Post-processes the decoded text before returning it.
+
+        This method is called after decoding and after caching.
+        It is useful to make this class more modular as some tokenizers need
+        special post-processing.
+        """
+        return text
+
+    def tokenize(self, request: TokenizationRequest) -> TokenizationRequestResult:
+        """Tokenizes `request.text` using `request.tokenizer`.
+
+        This method handles caching and returning the appropriate object while the actual tokenization
+        logic lies in the `_get_tokenize_do_it` method.
+        Most tokenizers hould simply implement `_get_tokenize_do_it` and leave this method as is.
+        However in some cases, such as the AI21 tokenizer, the tokenization logic is more complex and
+        requires additional logic, so this method can be overridden.
+
+        Returns a `TokenizationRequestResult` object.
+        """
+        cache_key = {
+            "text": request.text,
+            "tokenizer": request.tokenizer,
+            "max_length": request.max_length if request.truncation else None,
+            "encode": request.encode if self.use_encode_in_cache_key else None,
+        }
```

Comment:
A couple problems with this design

1. The cache key changes, so we can't use our existing tokenizer cache.
2. There's a implicit expectations the structure on the response dict.

I think the method signatures should be:

```
@abstractmethod
def _tokenization_request_to_raw_request(self, request: TokenizationRequest) -> Dict[str, Any]:
   pass

@abstractmethod
def _tokenization_raw_response_to_text_and_tokens(self, raw_response: Dict[str, Any]) -> Tuple[str, List[TokenizationToken]:
   pass

@abstractmethod
def _decode_raw_response_to_text(self, raw_response: Dict[str, Any]) -> str:
   pass

@abstractmethod
def _tokenize_do_it(self, request: Dict[str, Any]) -> Dict[str, Any]:
   pass
```

Then this would look like:

```
cache_key = self._tokenization_request_to_raw_request(request)
raw_response, cached = self.cache.get(cache_key, wrap_request_time(self._tokenize_to_id))
text, tokens = self._raw_respones_to_text_and_tokens(raw_response)
return TokenizationRequestResult(
    success=True,
    cached=cached,
    text=text,
    tokens=tokens,
    request_time=response["request_time"],
    error=None,
)
```

where the raw request and response formats are preserved from the existing clients.

Then you can also get rid of `_post_process_tokenization` and `_post_process_decoding`.

Reply:
Well this means reincluding a lot of the complexity for no reason except preserving the existing cache.
For example, some raw responses (before this PR) return these keys: `"tokens", "token_strings", "token_ids", ...`. The idea was to unify all of this.

Regarding your second point I am not sure I understand. Should we make an Object `TokenizationRawResponse` if this is what bothers you?

Reply:
The main intent is that the cache should preserve _as close as possible to the exact thing_ that gets sent to the tokenizer API and is returned from the tokenizer API.

Reply:
Done